### PR TITLE
163 move to cloudinary

### PR DESCRIPTION
**Pull request by kevinpmcc**
## What this does
- [x] replaces paperclip with carrierwave gem for image uploading
- [x] replaces AWS with cloudinary for image storage.

This is my first qa pull request on project so any feedback about style, practices etc. would be very much appreciated.

The data around image_content_type, image_file_size, and image_updated_at in the drawing table were being added automatically through paperclip i think. They are not currently being filled.  I can manually create these as part of drawing object as the file is uploaded if we want. Could you let me know if this data is being used?

I have no tests currently for this, need to investigate more exactly what need to tested without uploading to cloudinary.

## Screenshots
hopefully all the same for users!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -174,12 +174,12 @@ Or for just one set of tests:
 
 ### Deployment and Production Notes
 
-Draw My Life is currently hosted on [Heroku](https://www.heroku.com/), with images automagically hosted on a free tier [Amazon Simple Storage Service (S3)](https://aws.amazon.com/documentation/s3/) instance. The `.env.production.example` file holds a template of the ENV vars required to be set on the live production environment.
+Draw My Life is currently hosted on [Heroku](https://www.heroku.com/), with images automagically hosted on [Cloudinary](http://www.cloudinary.com/) instance. The `.env.production.example` file holds a template of the ENV vars required to be set on the live production environment.
```
Comment (krissy) [inline README.md]: 5 ⭐️ for remembering to update this! One small edit, we can remove the "instance" after automatically hosted on ....

File: app/models/drawing.rb [hunk 2]
```diff
@@ -16,16 +16,17 @@ class Drawing < ActiveRecord::Base
...
   validates :mood_rating, numericality: { only_integer: true, greater_than_or_equal_to: 1, less_than_or_equal_to: 5 }, allow_nil: true
 
-  has_attached_file :image, styles: {
-    medium: ["640x", :jpg],
-    thumb: ["100x100#", :jpg],
-    large: ["100%", :jpg]
-  }, convert_options: {
-    all: "-bordercolor none -border 1 -trim"
-  }
+  # has_attached_file :image, styles: {
```
Comment (krissy) [inline app/models/drawing.rb]: Assume you left some lines commented out to confirm we don't need them anymore.

This was specifically to render sizes for Paperclip so we can remove this altogether :+1:

File: app/models/drawing.rb [hunk 3]
```diff
@@ -16,16 +16,17 @@ class Drawing < ActiveRecord::Base
...
+  #  medium: ["640x", :jpg],
+  #  thumb: ["100x100#", :jpg],
+  #  large: ["100%", :jpg]
+  # }, convert_options: {
+  #  all: "-bordercolor none -border 1 -trim"
+  # }
 
-  validates_attachment_content_type :image, content_type: %r{\A(image\/(jpeg|png|gif|tiff|bmp)|application\/pdf)\z},
-                                            message: "Accepted image formats are: jpg/jpeg, png, tiff, gif, bmp, pdf"
+  mount_uploader :image, ImageUploader
```
Comment (krissy) [inline app/models/drawing.rb]: Can we create a custom validation for this via CarrierWave / Cloudinary? Looks like unlike Paperclip there's no built-in validator, but we could do something like this but for accepted content formats? https://github.com/carrierwaveuploader/carrierwave/wiki/How-to%3A-Validate-image-file-size

File: app/uploaders/image_uploader.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
+class ImageUploader < CarrierWave::Uploader::Base
```
Comment (krissy) [inline app/uploaders/image_uploader.rb:1]: What you've enabled / disabled looks good to me, so can remove the commented code altogether (minus descriptive comments)

File: config/environments/production.rb [hunk 5]
```diff
@@ -87,16 +87,16 @@
...
-  config.paperclip_defaults = {
-    storage: :s3,
-    s3_credentials: {
-      bucket: ENV.fetch('S3_BUCKET_NAME', ''),
-      access_key_id: ENV.fetch('AWS_ACCESS_KEY_ID', ''),
-      secret_access_key: ENV.fetch('AWS_SECRET_ACCESS_KEY', '')
-    },
-    s3_protocol: :https
-  }
+  # # Paperclip Amazon S3 hosting
```
Comment (krissy) [inline config/environments/production.rb]: We can remove this, version history will save us if it gets to having to reintroduce it! (I'm sure it won't :))

File: config/initializers/cloudinary.rb [hunk 6]
```diff
@@ -0,0 +1,3 @@
+Cloudinary.config do |config|
+  config.secure = true
```
Comment (krissy) [inline config/initializers/cloudinary.rb:2]: :+1:

File: config/environments/test.rb [hunk 7]
```diff
@@ -41,7 +41,7 @@
   # config.action_view.raise_on_missing_translations = true
 
   # Prevent multiple copies of Paperclip test files with a custom path
-  Paperclip::Attachment.default_options[:path] = "#{Rails.root}/spec/test_files/:class/:id_partition/:style.:extension"
+  # Paperclip::Attachment.default_options[:path] = "#{Rails.root}/spec/test_files/:class/:id_partition/:style.:extension"
```
Comment (krissy) [inline config/environments/test.rb]: Ditto, can remove this

File: lib/tasks/backfill_drawing_mood_rating_field.rake [hunk 8]
```diff
@@ -1,7 +1,6 @@
 namespace :backfill do
-
   desc "Halves mood rating of any images uploaded before 24 Oct 16"
-  task :mood_rating_field => :environment do
+  task mood_rating_field: 'environment' do
```
Comment (krissy) [inline lib/tasks/backfill_drawing_mood_rating_field.rake:3]: :+1:

File: spec/models/drawing_spec.rb [hunk 9]
```diff
@@ -27,18 +27,6 @@
         .is_less_than_or_equal_to(5)
     end
 
-    context "images/paperclip" do
```
Comment (krissy) [inline spec/models/drawing_spec.rb:30]: As you mentioned, will be nice to replace these once we figure out validation in Carrierwave!

File: spec/rails_helper.rb [hunk 10]
```diff
@@ -8,7 +8,7 @@
 require 'rspec/its'
 # Add additional requires below this line. Rails is not loaded until this point!
 require 'devise'
-require "paperclip/matchers"
+# require "paperclip/matchers"
```
Comment (krissy) [inline spec/rails_helper.rb]: Feel free to remove

**Review by krissy — COMMENTED** [review]
Changes so far look good! Yes go ahead and remove the ` image_content_type, image_file_size, and image_updated_at` fields - they're not being used. 

And as discussed offline, this will be blocked from being merged until we run a separate s3 -> cloudinary migration script for existing images which you're working on now.

File: app/models/drawing.rb [hunk 11]
```diff
@@ -16,16 +17,8 @@ class Drawing < ActiveRecord::Base
...
-    thumb: ["100x100#", :jpg],
-    large: ["100%", :jpg]
-  }, convert_options: {
-    all: "-bordercolor none -border 1 -trim"
-  }
-
-  validates_attachment_content_type :image, content_type: %r{\A(image\/(jpeg|png|gif|tiff|bmp)|application\/pdf)\z},
-                                            message: "Accepted image formats are: jpg/jpeg, png, tiff, gif, bmp, pdf"
+  mount_uploader :image, ImageUploader
+  validates :image, presence: true
```
Comment (kevinpmcc) [inline app/models/drawing.rb]: by moving validate :image, presence: true below mount_uploader the extension_whitelist validaton error is now displayed. (although it didn't actually exist so I added it to en.yml file

**Comment by kevinpmcc** [comment]
updated all based on comments, thanks @krissy 
validations now working by moving 
<code>validate :image, presence: true </code> below <code>mount_uploader</code>
 the extension_whitelist validation error is now displayed. this didn't exist so is added to en.yml file.

File: app/uploaders/image_uploader.rb [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  # Create different versions of your uploaded files:
+  version :thumb do
+    process resize_to_fit: [100, 100]
+    process convert: [:jpg]
+  end
+
+  version :medium do
+    process resize_to_fit: [640, 640]
```
Comment (steverabino) [inline app/uploaders/image_uploader.rb]: We don't want this size to be 640x640, more like we want the width to be 640, but the height to be proportional to this

**Pull request closed without merging**